src/transformers/models/pop2piano/modeling_pop2piano.py

```diff
@@ -0,0 +1,1707 @@
+                passed to generate matching the attributes of `generation_config` will override them. If
+                `generation_config` is not provided, the default will be used, which had the following loading
+                priority: 1) from the `generation_config.json` model file, if it exists; 2) from the model
+                configuration. Please note that unspecified parameters will inherit [`~generation.GenerationConfig`]'s
+                default values, whose documentation should be checked to parameterize generation.
+            return_dict_in_generate (`bool`, *optional*, defaults to `True`):
+                Whether to return dict or not.
+            kwargs:
+                Ad hoc parametrization of `generate_config` and/or additional model-specific kwargs that will be
+                forwarded to the `forward` function of the model. If the model is an encoder-decoder model, encoder
+                specific kwargs should not be prefixed and decoder specific kwargs should be prefixed with *decoder_*.
+        Return:
+            [`~utils.ModelOutput`] or `torch.LongTensor`: A [`~utils.ModelOutput`] (if `return_dict_in_generate=True`
+            or when `config.return_dict_in_generate=True`) or a `torch.FloatTensor`.
+                Since Pop2Piano is an encoder-decoder model (`model.config.is_encoder_decoder=True`), the possible
+                [`~utils.ModelOutput`] types are:
+                    - [`Pop2PianoGreedySearchEncoderDecoderOutput`],
+                    - [`Pop2PianoSampleEncoderDecoderOutput`],
+                    - [`Pop2PianoBeamSearchEncoderDecoderOutput`],
+                    - [`Pop2PianoBeamSampleEncoderDecoderOutput`]
+        """
+
+        if generation_config is None:
+            generation_config = self.generation_config
+        generation_config.update(**kwargs)
+
+        # check for composer_to_feature_token
+        if not hasattr(generation_config, "composer_to_feature_token"):
+            raise ValueError(
+                "`composer_to_feature_token` was not found! Please refer to "
+                "https://huggingface.co/susnato/pop2piano_dev/blob/main/generation_config.json"
+                "and parse a dict like that."
+            )
+        else:
+            if len(generation_config.composer_to_feature_token) != self.config.composer_vocab_size:
+                raise ValueError(
+                    "config.composer_vocab_size must be same as the number of keys in "
+                    f"generation_config.composer_to_feature_token! "
+                    f"Found {self.config.composer_vocab_size} vs {len(generation_config.composer_to_feature_token)}."
+                )
+
+        # check for attention_mask
+        if input_features["input_features"].shape[0] > 1 and not hasattr(
+            input_features, "attention_mask_input_features"
+        ):
+            raise ValueError("attention_mask must be present for batched inputs!")
```

Comment:
After this line we should be able to directly call `super.generate`? Don't think there's a need to re-write the generation loop. We can modify the generation signature as:
```python
    @torch.no_grad()
    def generate(
        self,
        input_features: torch.Tensor,
        attention_mask: torch.Tensor=None,
        composer="composer1",
        generation_config=None,
        **kwargs,
    ):
```
(note that I've removed max length and return dict in generate, these can form part of the kwargs) And then simply do:
```python
super.generate(inputs=input_features, attention_mask=attention_mask, generation_config=generation_config, **kwargs)
```

The generate method expects an attention mask anyway: https://github.com/huggingface/transformers/blob/6b7d6f848bffb2ed68935858e71395c4070e3200/src/transformers/generation/utils.py#L1265

We should make use of it!

Reply:
Since the attention_mask for input_features is of shape (dim1,dim2,dim3,dim4)(eg. [2, 120, 66, 512]), I don't think it will be compatible with `super.generate` and thanks for pointing out the `max_length` and `return_dict_in_generate` I will definitely remove it in the next commit. 

Please let me know if these reasons I mentioned are ok or not and if I need to change them.

Reply:
All generate is doing is calling 1 forward call of the encoder (to get the encoder hidden states), and then N forward calls of the decoder (to get the predicted tokens)

Reply:
Here the `Pop2PianoFeatureExtractor` generates input_embeds of shape
 (`batch_size`, `timestamp`, `sequence_len`, `embed_dim`) and the shape of attention_mask is (`batch_size`, `max_timestamp`, `max_sequence_len`, `embed_dim`). Since `timestamp` and `sequence_len` are  different for every audio(because of `extract_rhythm` as I said earlier) the `Pop2PianoFeatureExtractor` pads every axis to the max value of that axis among all audios.
Thats why the attention_mask provided by `Pop2PianoFeatureExtractor` is not compatible with the forward signature of `Pop2PianoForConditionalGeneration` but it is compatible with the `generate` as in generate we use a for loop(https://github.com/susnato/transformers/blob/290411987506b01b7fd3294ec074e228491f408a/src/transformers/models/pop2piano/modeling_pop2piano.py#L1556) and inside that loop, for each audio_embed we extract the real values from padded ones and then feed them in generate as input_embeds. 

Actually we can't directly pass the values as input_embeds in `super.generate`, since the shapes won't match, we need to take each audio embed and then manually pass them as we are doing in the loop.